File: Day-4-linked-list/devops-usecase/golang/command_list.go
package main

import (
	"bytes"
	"fmt"
	"os/exec"
	"strings"
)

type CommandNode struct {
	command string
	status  *bool // nil: not executed, true: success, false: failure
	output  string
	error   string
	prev    *CommandNode
	next    *CommandNode
}

type CommandList struct {
	head *CommandNode
}

func NewCommandList() *CommandList {
	return &CommandList{}
}

func (cl *CommandList) AddCommand(command string) {
	newNode := &CommandNode{command: command}
	if cl.head == nil {
		cl.head = newNode
	} else {
		temp := cl.head
		for temp.next != nil {
			temp = temp.next
		}
		temp.next = newNode
		newNode.prev = temp
	}
}

func (cl *CommandList) ExecuteCommands() {
	temp := cl.head
	for temp != nil {
		cmd := exec.Command("sh", "-c", temp.command)
		var out bytes.Buffer
		var stderr bytes.Buffer
		cmd.Stdout = &out
		cmd.Stderr = &stderr
		err := cmd.Run()
		if err != nil {
			status := false
			temp.status = &status
			temp.error = stderr.String()
			fmt.Printf("Command '%s' failed with error: %s\n", temp.command, temp.error)
		} else {
			status := true
			temp.status = &status
			temp.output = out.String()
			fmt.Printf("Command '%s' succeeded.\n", temp.command)
		}
		temp = temp.next
	}
}

func (cl *CommandList) PrintCommandsStatus() {
	temp := cl.head
	for temp != nil {
		status := "Not Executed"
		if temp.status != nil {
			if *temp.status {
				status = "Success"
			} else {
				status = "Failure"
			}
		}
		fmt.Printf("Command: %s, Status: %s\n", temp.command, status)
		if temp.output != "" {
			fmt.Printf("Output: %s\n", strings.TrimSpace(temp.output))
		}
		if temp.error != "" {
			fmt.Printf("Error: %s\n", strings.TrimSpace(temp.error))
		}
		fmt.Println()
		temp = temp.next
	}
}

func (cl *CommandList) PrintCommandsForward() {
	temp := cl.head
	for temp != nil {
		fmt.Print(temp.command)
		if temp.next != nil {
			fmt.Print(" <-> ")
		}
		temp = temp.next
	}
	fmt.Println()
}

func (cl *CommandList) PrintCommandsBackward() {
	temp := cl.head
	if temp == nil {
		return
	}
	for temp.next != nil {
		temp = temp.next
	}
	for temp != nil {
		fmt.Print(temp.command)
		if temp.prev != nil {
			fmt.Print(" <-> ")
		}
		temp = temp.prev
	}
	fmt.Println()
}

func main() {
	cmdList := NewCommandList()
	cmdList.AddCommand("echo 'Hello World'")
	cmdList.AddCommand("ls -l")
	cmdList.AddCommand("invalid_command") // This will simulate a failed command
	cmdList.AddCommand("pwd")

	fmt.Println("Commands in order (Forward):")
	cmdList.PrintCommandsForward()

	fmt.Println("\nExecuting commands...")
	cmdList.ExecuteCommands()

	fmt.Println("\nCommands status:")
	cmdList.PrintCommandsStatus()

	fmt.Println("\nCommands in order (Backward):")
	cmdList.PrintCommandsBackward()
}

File: Day-4-linked-list/devops-usecase/python/command_list.py
import subprocess

class CommandNode:
    def __init__(self, command):
        self.command = command
        self.status = None  # None: not executed, True: success, False: failure
        self.output = None
        self.error = None
        self.prev = None
        self.next = None

class CommandList:
    def __init__(self):
        self.head = None

    def add_command(self, command):
        # Create a new command node with the given command
        new_node = CommandNode(command)
        # If the list is empty, initialize the head to the new node
        if not self.head:
            self.head = new_node
        else:
            # Traverse to the end of the list
            temp = self.head
            while temp.next:
                temp = temp.next
            # Insert the new node at the end and update pointers
            temp.next = new_node
            new_node.prev = temp

    def execute_commands(self):
        temp = self.head
        # Traverse through the list and execute each command
        while temp:
            try:
                # Run the command using subprocess and capture the output
                result = subprocess.run(temp.command, shell=True, text=True, capture_output=True, check=True)
                temp.status = True
                temp.output = result.stdout
                temp.error = result.stderr
                print(f"Command '{temp.command}' succeeded.")
            except subprocess.CalledProcessError as e:
                temp.status = False
                temp.output = e.stdout
                temp.error = e.stderr
                print(f"Command '{temp.command}' failed with error: {e.stderr.strip()}")
            temp = temp.next

    def print_commands_status(self):
        temp = self.head
        # Traverse through the list and print the status of each command
        while temp:
            status = "Success" if temp.status else "Failure" if temp.status is not None else "Not Executed"
            print(f"Command: {temp.command}, Status: {status}")
            if temp.output:
                print(f"Output: {temp.output.strip()}")
            if temp.error:
                print(f"Error: {temp.error.strip()}")
            print()
            temp = temp.next

    def print_commands_forward(self):
        temp = self.head
        # Traverse through the list and print each command in forward order
        while temp:
            print(temp.command, end=' <-> ' if temp.next else '')
            temp = temp.next
        print()

    def print_commands_backward(self):
        temp = self.head
        # Traverse to the end of the list
        while temp and temp.next:
            temp = temp.next
        # Traverse backward through the list and print each command
        while temp:
            print(temp.command, end=' <-> ' if temp.prev else '')
            temp = temp.prev
        print()

# Usage
cmd_list = CommandList()
cmd_list.add_command("echo 'Hello World'")
cmd_list.add_command("ls -l")
cmd_list.add_command("invalid_command")  # This will simulate a failed command
cmd_list.add_command("pwd")

print("Commands in order (Forward):")
cmd_list.print_commands_forward()

print("\nExecuting commands...")
cmd_list.execute_commands()

print("\nCommands status:")
cmd_list.print_commands_status()

print("\nCommands in order (Backward):")
cmd_list.print_commands_backward()
